src/common/lib/python/common_extension.cc

```diff
@@ -404,63 +461,128 @@ static int PyTask_init(PyTask *self, PyObject *args, PyObject *kwds) {
 }
 
 static void PyTask_dealloc(PyTask *self) {
-  if (self->spec != NULL) {
+  if (!use_raylet(self)) {
     TaskSpec_free(self->spec);
+  } else {
+    delete self->task_spec;
   }
   delete self->execution_dependencies;
-  Py_TYPE(self)->tp_free((PyObject *) self);
+  Py_TYPE(self)->tp_free(reinterpret_cast<PyObject *>(self));
 }
 
-static PyObject *PyTask_function_id(PyObject *self) {
-  FunctionID function_id = TaskSpec_function(((PyTask *) self)->spec);
+static PyObject *PyTask_function_id(PyTask *self) {
+  FunctionID function_id = TaskSpec_function(self->spec);
```

Comment:
Doesn't this also need the branch on `use_raylet`?

Reply:
Good catch, I'm surprised things worked without this..